references/cli/dryrun.go

```diff
@@ -401,3 +403,64 @@ func getPolicyNameFromWorkflow(wf *wfv1alpha1.Workflow, policyNameMap map[string
 	}
 	return nil
 }
+
+// includeBuiltinWorkflowStepDefinition adds builtin workflow step definition to the given objects
+// A few builtin workflow steps have cue definition. They should be included when building offline fake client.
+func includeBuiltinWorkflowStepDefinition(objs []*unstructured.Unstructured) []*unstructured.Unstructured {
+	deployUnstructured, _ := oamutil.Object2Unstructured(deployDefinition)
+	return append(objs, deployUnstructured)
+}
+
+// deployDefinition is the definition of deploy step
+// Copied it here to make dry-run work in offline mode.
+var deployDefinition = &corev1beta1.WorkflowStepDefinition{
+	TypeMeta: metav1.TypeMeta{
+		Kind:       corev1beta1.WorkflowStepDefinitionKind,
+		APIVersion: corev1beta1.SchemeGroupVersion.String(),
+	},
+	ObjectMeta: metav1.ObjectMeta{
+		Name:      "deploy",
+		Namespace: oam.SystemDefinitionNamespace,
+	},
+	Spec: corev1beta1.WorkflowStepDefinitionSpec{
+		Schematic: &apicommon.Schematic{
+			CUE: &apicommon.CUE{Template: `
+import (
+	"vela/op"
+)
+
+"deploy": {
+	type: "workflow-step"
+	annotations: {
+		"category": "Application Delivery"
+	}
+	labels: {
+		"scope": "Application"
+	}
+	description: "A powerful and unified deploy step for components multi-cluster delivery with policies."
+}
+template: {
+	if parameter.auto == false {
+		suspend: op.#Suspend & {message: "Waiting approval to the deploy step \"\(context.stepName)\""}
+	}
+	deploy: op.#Deploy & {
+		policies:                 parameter.policies
+		parallelism:              parameter.parallelism
+		ignoreTerraformComponent: parameter.ignoreTerraformComponent
+	}
+	parameter: {
+		//+usage=If set to false, the workflow will suspend automatically before this step, default to be true.
+		auto: *true | bool
+		//+usage=Declare the policies that used for this deployment. If not specified, the components will be deployed to the hub cluster.
+		policies: *[] | [...string]
+		//+usage=Maximum number of concurrent delivered components.
+		parallelism: *5 | int
+		//+usage=If set false, this step will apply the components with the terraform workload.
+		ignoreTerraformComponent: *true | bool
+	}
+}
+`,
+			},
+		},
+	},
```

Comment:
what if the CUE files changed in the future? Can we just leave the deploy step only with no parameters?

```
"deploy": {
	type: "workflow-step"
	annotations: {
		"category": "Application Delivery"
	}
	labels: {
		"scope": "Application"
	}
	description: "This is an inner place holder for built-in deploy step."
}
template: {
}
```

Reply:
Template removed.